From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tatsuyuki Ishi <ishitatsuyuki@gmail.com>
Date: Fri, 13 Oct 2023 16:46:09 +0000
Subject: [PATCH] connectivity: Make curl timeout callback non-repeating.

The CURLMOPT_TIMERFUNCTION manual says:
Your callback function timer_callback should install a *non-repeating* timer with an expire time of timeout_ms milliseconds.

The previous callback returned G_SOURCE_CONTINUE, resulting in a repeating firing callback. curl tweaked some internals in v8.4.0 and started giving 0 timeouts, which caused this to wreak havoc.

Fix it by making the callback non-repeating as specified by the API contract.
---
 src/core/nm-connectivity.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/core/nm-connectivity.c b/src/core/nm-connectivity.c
index 92de44f17d99..2a34752558e8 100644
--- a/src/core/nm-connectivity.c
+++ b/src/core/nm-connectivity.c
@@ -408,7 +408,7 @@ _con_curl_timeout_cb(gpointer user_data)
 
     _con_curl_check_connectivity(cb_data->concheck.curl_mhandle, CURL_SOCKET_TIMEOUT, 0);
     _complete_queued(cb_data->self);
-    return G_SOURCE_CONTINUE;
+    return G_SOURCE_REMOVE;
 }
 
 static int
